### DISCUSSION TITLE
Upgrade and rollback strategy

### DISCUSSION THREAD

**Discussion by ashep** [zitadel/zitadel#8711] [Q&A] [answered]
Hello!

I have an on-premises Zitadel installation, and I wonder whether Zitadel guarantees any forward compatibility.

For example, I have upgraded from version 2.58.8 to 2.63.0 (version numbers have been taken arbitrarily),  and after the upgrading, I've noticed performance degradation or anything else.

The question is, is it safe to revert the installation to the previously installed version **without rolling back the database**?

I appreciate any additional information related to this question, namely, how to upgrade Zitadel in the safest way.

Thank you!

**Comment by hifabienne** [comment]
@adlerhurst @muhlemmer can you help here?

**Comment by muhlemmer** [comment, marked as answer]
Hi @ashep. Yes, in general it is safe to rollback. We have an event-driven model. So when database models ("projections") change, they are recreated using the events from the event store on upgrade. We use versioned projections, so different version of zitadel may use different version of the projections. For example for the users projection:

- ZITADEL v2.54 uses a table called `projections.users12`
- ZITADEL v2.55 and later uses a table called `projections.users13`

From a database perspective it's even possible to run 2 different version of zitadel in parallel behind a loadbalancer. If there are different tables, they each use their own version of projection tables.

Only when we do change something in the eventstore tables it becomes tricky. In such case we inform the users about a migration through a [technical advisory](https://zitadel.com/docs/support/technical_advisory).

**Reply by ashep** [reply]
Thank you for the detailed explanation, @muhlemmer!

Thenk you for the help, @hifabienne!

Have a good day ;)
